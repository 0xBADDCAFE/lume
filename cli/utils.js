import { existsSync } from "../deps/fs.js";
import lume from "../mod.js";
import { join, resolve } from "../deps/path.js";
import { brightGreen } from "../deps/colors.js";

/**
 * @return {Promise<*>} a lume instance - ready to build, run, etc.
 */
export async function buildSite(options, site) {
  options.root = resolve(Deno.cwd(), options.root);
  const configFile = join(options.root, options.config);

  if (!site) {
    if (existsSync(configFile)) {
      const mod = await import(`file://${configFile}`);
      site = mod.default;
    } else {
      site = lume();
    }
  }

  site.options.cwd = options.root;

  if (options.dev) {
    site.options.dev = options.dev;
  }

  if (options.location) {
    site.options.location = new URL(options.location);
  }

  if (options.src) {
    site.options.src = options.src;
  }

  if (options.dest) {
    site.options.dest = options.dest;
  }

  if (options["--"]) {
    site.options.flags = options["--"];
  }

  return site;
}

/**
 * @param command name of the command you're validating
 * @param args array of cli arguments (i.e. options._)
 * @param max the max number of args
 * @param min the min number of args
 */
export function validateArgsCount(command, args, max, min = 0) {
  if (args.length > max) {
    throw new Error(`Unexpected arguments: ${args.slice(max).join(", ")}
    
    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }

  if (args.length < min) {
    throw new Error(`Missing arguments

    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }
}
