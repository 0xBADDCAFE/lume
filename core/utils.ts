import { DOMParser, HTMLDocument } from "../deps/dom.ts";
import {
  brightGreen,
  brightYellow,
  cyan,
  dim,
  green,
  red,
} from "../deps/colors.ts";
import { dirname, extname, join, posix, SEP } from "../deps/path.ts";
import { parse } from "../deps/jsonc.ts";
import { Exception } from "./errors.ts";

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "code_highlight",
  "date",
  "esbuild",
  "eta",
  "imagick",
  "inline",
  "jsx",
  "jsx_preact",
  "katex",
  "lightningcss",
  "liquid",
  "metas",
  "minify_html",
  "modify_urls",
  "multilanguage",
  "netlify_cms",
  "on_demand",
  "pagefind",
  "postcss",
  "prism",
  "pug",
  "relations",
  "relative_urls",
  "remark",
  "resolve_urls",
  "sass",
  "sheets",
  "slugify_urls",
  "svgo",
  "terser",
  "windi_css",
];

/** A list of the available plugins with init configurations */
export const initPlugins = [
  "jsx",
  "jsx_preact",
];

export function log(...lines: (string | undefined)[]) {
  console.log("----------------------------------------");
  lines.forEach((line) => line && console.log(line));
  console.log("----------------------------------------");
}

export function promptConfigUpdate(data: unknown) {
  log(
    red("deno.jsonc needs to be manually updated:"),
    dim("Use deno.json to update it automatically"),
    JSON.stringify(data, null, 2),
  );
}

/** Check the compatibility with the current Deno version */
export function checkDenoVersion(): void {
  const minimum = "1.25.4";
  const current = Deno.version.deno;

  if (current < minimum) {
    console.log("----------------------------------------");
    console.error(red("Your Deno version is not compatible with Lume"));
    console.log(`Lume needs Deno ${green(minimum)} or greater`);
    console.log(`Your current version is ${red(current)}`);
    console.log(`Run ${cyan("deno upgrade")} and try again`);
    console.log("----------------------------------------");
    Deno.exit(1);
  }
}

/** Check if the current version is outdated */
export async function checkUpgrade(): Promise<void> {
  const current = getLumeVersion();

  // It's a local version
  if (current.startsWith("local ")) {
    return;
  }

  const stable = !!current.match(/^v\d+\./);
  const expires = 1000 * 60 * 60 * 24; // 1 day
  const interval = localStorage.getItem("lume-upgrade");

  if (interval && parseInt(interval) + expires > Date.now()) {
    return;
  }

  localStorage.setItem("lume-upgrade", Date.now().toString());

  const latest = stable
    ? await getLatestVersion()
    : await getLatestDevelopmentVersion();

  if (current === latest) {
    return;
  }

  let global = false;
  try {
    await Promise.any([
      Deno.stat("deno.json"),
      Deno.stat("deno.jsonc"),
    ]);
    await Deno.stat("import_map.json");
  } catch {
    global = true;
  }

  const command = global
    ? (stable ? "lume upgrade --global" : "lume upgrade --dev --global")
    : (stable ? "deno task lume upgrade" : "deno task lume upgrade --dev");

  log(
    `Update available ${dim(current)}  â†’ ${green(latest)}`,
    `Run ${cyan(command)} to update`,
  );
}

/** Return the latest stable version from the deno.land/x repository */
export async function getLatestVersion(): Promise<string> {
  const response = await fetch("https://cdn.deno.land/lume/meta/versions.json");
  const versions = await response.json();
  return versions.latest;
}

/** Return the hash of the latest commit from the GitHub repository */
export async function getLatestDevelopmentVersion(): Promise<string> {
  const response = await fetch(
    "https://api.github.com/repos/lumeland/lume/commits?per_page=1",
  );
  const commits = await response.json();
  return commits[0].sha;
}

/** Import map file */
export interface ImportMap {
  imports: Record<string, string>;
  scopes?: Record<string, Record<string, string>>;
}

/** Basic options for deno.json file */
export interface DenoConfig {
  importMap?: string;
  tasks?: Record<string, string>;
  compilerOptions?: {
    jsx?: "jsx" | "react-jsx";
    jsxImportSource?: string;
  };
  [key: string]: unknown;
}

/** Run a callback concurrently with all the elements of an Iterable */
export async function concurrent<Type>(
  iterable: AsyncIterable<Type> | Iterable<Type>,
  iteratorFn: (arg: Type) => Promise<unknown>,
  limit = 200,
) {
  const executing: Promise<unknown>[] = [];

  for await (const item of iterable) {
    const p: Promise<unknown> = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

const decoder = new TextDecoder();
const encoder = new TextEncoder();

/** Encode a message using SHA-1 algorithm */
export async function sha1(message: string | Uint8Array): Promise<string> {
  if (typeof message === "string") {
    message = encoder.encode(message);
  }

  const hash = await crypto.subtle.digest("SHA-1", message);
  return decoder.decode(hash);
}

/** Helper to create optional properties recursively */
// deno-lint-ignore ban-types
export type DeepPartial<T> = T extends object ? {
    [P in keyof T]?: DeepPartial<T[P]>;
  }
  : T;

/**
 * Merge two objects recursively.
 * It's used to merge user options with default options.
 */
export function merge<Type>(
  defaults: Type,
  user?: Partial<Type> | DeepPartial<Type>,
): Type {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    if (value === undefined) {
      continue;
    }

    // @ts-ignore: No index signature with a parameter of type 'string' was found on type 'unknown'
    if (isPlainObject(merged[key]) && isPlainObject(value)) {
      // @ts-ignore: Type 'string' cannot be used to index type 'Type'
      merged[key] = merge(merged[key], value);
      continue;
    }

    // @ts-ignore: Type 'string' cannot be used to index type 'Type'
    merged[key] = value;
  }

  return merged;
}

const reactElement = Symbol.for("react.element");
const objectConstructor = {}.constructor;

/** Check if the argument passed is a plain object */
export function isPlainObject(obj: unknown): obj is Record<string, unknown> {
  return typeof obj === "object" && obj !== null &&
    obj.constructor === objectConstructor &&
    // @ts-ignore: Check if the argument passed is a React element
    obj["$$typeof"] !== reactElement;
}

/**
 * Convert the Windows paths (that use the separator "\")
 * to Posix paths (with the separator "/")
 * and ensure it starts with "/".
 */
export function normalizePath(path: string) {
  if (SEP !== "/") {
    path = path.replaceAll(SEP, "/");

    // Is absolute Windows path (C:/...)
    if (path.includes(":/")) {
      return path;
    }
  }

  return posix.join("/", path);
}

/** Convert an HTMLDocument instance to a string */
export function documentToString(document: HTMLDocument) {
  const { doctype, documentElement } = document;

  if (!doctype) {
    return `<!DOCTYPE html>\n${documentElement?.outerHTML || ""}`;
  }

  return `<!DOCTYPE ${doctype.name}` +
    (doctype.publicId ? ` PUBLIC "${doctype.publicId}"` : "") +
    (!doctype.publicId && doctype.systemId ? " SYSTEM" : "") +
    (doctype.systemId ? ` "${doctype.systemId}"` : "") +
    `>\n${documentElement?.outerHTML}`;
}

const parser = new DOMParser();

/** Parse a string with HTML code and return an HTMLDocument */
export function stringToDocument(string: string): HTMLDocument {
  const document = parser.parseFromString(string, "text/html");

  if (!document) {
    throw new Error("Unable to parse the HTML code");
  }

  return document;
}

/** Return the current installed version */
export function getLumeVersion(
  url = new URL(import.meta.resolve("../")),
): string {
  const { pathname } = url;
  return pathname.match(/@([^/]+)/)?.[1] ?? `local (${pathname})`;
}

/** Returns the _config file of a site */
export async function getConfigFile(
  path?: string,
): Promise<string | undefined> {
  if (path) {
    try {
      return await Deno.realPath(path);
    } catch (cause) {
      throw new Exception("Config file not found", { path, cause });
    }
  }

  const paths = ["_config.js", "_config.ts"];

  for (const path of paths) {
    try {
      return await Deno.realPath(path);
    } catch {
      // Ignore
    }
  }
}

export interface DenoConfigResult {
  file: string;
  config: DenoConfig;
  importMap?: ImportMap;
}

/** Detect and returns the Deno configuration */
export async function readDenoConfig(
  importMapFile?: string,
): Promise<DenoConfigResult | undefined> {
  for (const file of ["deno.json", "deno.jsonc"]) {
    try {
      const content = await Deno.readTextFile(file);
      const config = parse(content) as DenoConfig;
      importMapFile ||= config.importMap;
      if (importMapFile) {
        config.importMap = importMapFile;
        const importMap: ImportMap = isUrl(importMapFile)
          ? await (await fetch(importMapFile)).json()
          : await JSON.parse(await Deno.readTextFile(importMapFile));

        return { file, config, importMap };
      }
      return { file, config };
    } catch (err) {
      if (err instanceof Deno.errors.NotFound) {
        continue;
      }

      throw err;
    }
  }
}

/** Update the Deno configuration */
export async function writeDenoConfig(options: DenoConfigResult) {
  const { file, config, importMap } = options;

  if (importMap && !config.importMap) {
    config.importMap = "./import_map.json";
  }

  if (config.importMap) {
    const importMapFile = join(dirname(file), config.importMap);
    console.log(importMapFile);

    await Deno.writeTextFile(
      importMapFile,
      JSON.stringify(importMap, null, 2) + "\n",
    );
    console.log(brightGreen("Import map file saved:"), importMapFile);
  }

  if (extname(file) === ".jsonc") {
    console.log(
      brightYellow("deno.jsonc needs to be manually updated:"),
      "Use deno.json to update it automatically",
      JSON.stringify(config, null, 2),
    );
  } else {
    await Deno.writeTextFile(file, JSON.stringify(config, null, 2) + "\n");
    console.log(brightGreen("Deno configuration file saved:"), file);
  }
}

export function isUrl(path: string): boolean {
  return !!path.match(/^(https?|file):\/\//);
}

export function createDate(str: string | number): Date | undefined {
  if (typeof str === "number") {
    return new Date(str);
  }

  const datetime = str.match(
    /^(\d{4})-(\d\d)-(\d\d)(?:-(\d\d)-(\d\d)(?:-(\d\d))?)?$/,
  );

  if (datetime) {
    const [, year, month, day, hour, minute, second] = datetime;

    return new Date(Date.UTC(
      parseInt(year),
      parseInt(month) - 1,
      parseInt(day),
      hour ? parseInt(hour) : 0,
      minute ? parseInt(minute) : 0,
      second ? parseInt(second) : 0,
    ));
  }

  if (str.match(/^\d+$/)) {
    return new Date(parseInt(str));
  }
}

export async function read(path: string, isBinary: true): Promise<Uint8Array>;
export async function read(path: string, isBinary: false): Promise<string>;
export async function read(
  path: string,
  isBinary: boolean,
): Promise<string | Uint8Array> {
  if (!isUrl(path)) {
    return isBinary ? Deno.readFile(path) : Deno.readTextFile(path);
  }

  const response = await fetch(path);

  return isBinary
    ? new Uint8Array(await response.arrayBuffer())
    : response.text();
}
