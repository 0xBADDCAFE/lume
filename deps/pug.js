export * from "https://esm.sh/pug@3.0.0";
