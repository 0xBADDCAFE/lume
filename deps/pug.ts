export * from "https://cdn.jsdelivr.net/gh/lumeland/pug@v0.1.3/mod.ts";
