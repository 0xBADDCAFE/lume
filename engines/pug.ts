import Site from "../site.js";
import Engine from "./engine.ts";
import { Data, Helper, HelperOptions } from "../types.ts";

interface Options {
  filters: Record<string, Helper>;
}

interface PugEngine {
  compile: (
    input: string,
    options: Record<string, unknown>,
  ) => (data: Data) => string;
}

export default class Pug extends Engine {
  options: Options;
  engine: PugEngine;
  cache: Map<string, (data: Data) => string> = new Map();

  constructor(site: Site, engine: PugEngine, options: Options) {
    super(site);
    this.engine = engine;
    this.options = options;

    // Update cache
    site.addEventListener("beforeUpdate", () => this.cache.clear());
  }

  render(content: string, data: Data, filename: string) {
    if (!this.cache.has(filename)) {
      this.cache.set(
        filename,
        this.engine.compile(content, {
          ...this.options,
          filename,
        }),
      );
    }

    return this.cache.get(filename)!(data);
  }

  addHelper(name: string, fn: Helper, options: HelperOptions) {
    switch (options.type) {
      case "filter":
        this.options.filters ||= {};
        this.options.filters[name] = (text, opt) => {
          // @ts-ignore: opt is of type 'unknown'.
          delete opt.filename;
          // @ts-ignore: opt is of type 'unknown'.
          const args = Object.values(opt);
          return fn(text, ...args);
        };
        return;
    }
  }
}
