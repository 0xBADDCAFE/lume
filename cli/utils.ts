import lume from "../mod.ts";
import { toFileUrl } from "../deps/path.ts";
import { dim, red } from "../deps/colors.ts";
import { getConfigFile } from "../core/utils.ts";

import type { Site } from "../core.ts";

/** Create a site instance */
export async function createSite(root: string, config?: string): Promise<Site> {
  const path = await getConfigFile(root, config);

  if (path) {
    console.log(`Loading config file ${dim(path)}`);
    console.log();
    const mod = await import(toFileUrl(path).href);
    return mod.default;
  }

  return lume();
}

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "code_highlight",
  "date",
  "esbuild",
  "eta",
  "imagick",
  "inline",
  "jsx",
  "lightningcss",
  "liquid",
  "metas",
  "minify_html",
  "modify_urls",
  "multilanguage",
  "nano_jsx",
  "netlify_cms",
  "on_demand",
  "postcss",
  "prism",
  "pug",
  "relations",
  "relative_urls",
  "resolve_urls",
  "sass",
  "slugify_urls",
  "svgo",
  "terser",
  "windi_css",
];

export function log(...lines: (string | undefined)[]) {
  console.log("----------------------------------------");
  lines.forEach((line) => line && console.log(line));
  console.log("----------------------------------------");
}

export function promptConfigUpdate(data: unknown) {
  log(
    red("deno.jsonc needs to be manually updated:"),
    dim("Use deno.json to update it automatically"),
    JSON.stringify(data, null, 2),
  );
}
