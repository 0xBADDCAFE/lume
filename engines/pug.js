import * as pug from "../deps/pug.js";
import TemplateEngine from "./templateEngine.js";

export default class Pug extends TemplateEngine {
  cache = new Map();
  filters = {};

  //Update cache
  update(filenames) {
    for (const filename of filenames) {
      this.cache.delete(filename);
    }
  }

  render(content, data, filename) {
    if (!this.cache.has(filename)) {
      this.cache.set(
        filename,
        pug.compile(content, {
          filename,
          basedir: this.site.src("_includes"),
          filters: this.filters,
        }),
      );
    }

    return this.cache.get(filename)(data);
  }

  addFilter(name, fn) {
    this.filters[name] = fn;
  }
}
