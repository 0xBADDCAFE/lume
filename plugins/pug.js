import * as pug from "../deps/pug.ts";
import Pug from "../engines/pug.ts";
import loader from "../loaders/text.js";
import { merge } from "../utils.ts";

// Default options
const defaults = {
  extensions: [".pug"],
  includes: null,
  options: {},
};

export default function (userOptions) {
  return (site) => {
    const options = merge(
      { ...defaults, includes: site.includes() },
      userOptions,
    );

    options.options.basedir = options.includes;

    site.loadPages(
      options.extensions,
      loader,
      new Pug(site, pug, options.options),
    );
  };
}
