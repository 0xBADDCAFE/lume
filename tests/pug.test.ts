import { assertStrictEquals as equals } from "../deps/assert.ts";
import { build, getSite, testPage } from "./utils.ts";
import pug from "../plugins/pug.ts";

Deno.test("build a site with eta", async () => {
  const site = getSite({
    src: "pug",
    location: new URL("https://example.com/blog"),
  });

  site.use(pug());

  await build(site);

  testPage(site, "/extends", (page) => {
    equals(page.data.title, "Pug example");
    equals(page.data.url, "/extends/");
    equals(page.document?.querySelector("h1")?.innerText, "Home");
    equals(page.document?.querySelector("title")?.innerText, page.data.title);
  });

  testPage(site, "/filter", (page) => {
    equals(page.data.title, "Markdown content");
    equals(page.data.url, "/article.html");
    equals(page.document?.querySelector("h1")?.innerText, "This is a title");
    equals(page.document?.querySelectorAll("li")?.length, 2);
  });
});
