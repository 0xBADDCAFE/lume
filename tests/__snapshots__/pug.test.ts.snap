export const snapshot = {};

snapshot[`build a site with eta 1`] = `3`;

snapshot[`build a site with eta 2`] = `
{
  formats: [
    {
      engines: 1,
      ext: ".tmpl.ts",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".tmpl.js",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: undefined,
      ext: ".tmpl.json",
      pageLoader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: undefined,
      ext: ".json",
    },
    {
      engines: 1,
      ext: ".md",
      pageLoader: [AsyncFunction: text],
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
    },
    {
      componentLoader: [AsyncFunction: text],
      engines: 1,
      ext: ".njk",
      includesPath: "_includes",
      pageLoader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yaml",
      pageLoader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yml",
      pageLoader: [AsyncFunction: yaml],
    },
    {
      componentLoader: [AsyncFunction: text],
      engines: 1,
      ext: ".pug",
      includesPath: "_includes",
      pageLoader: [AsyncFunction: text],
    },
  ],
}
`;

snapshot[`build a site with eta 3`] = `
[
]
`;

snapshot[`build a site with eta 4`] = `
{
  content: '<html class="no-js" lang="en"><head><meta charset="utf-8"/><title>Pug example</title></head><body><h1>Home</h1></body></html>',
  data: {
    content: "extends /layout.pug
block content
  h1 Home
",
    date: 1970-01-01T00:00:00.000Z,
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    slug: "extends",
    tags: [
    ],
    title: "Pug example",
    url: "/extends/",
  },
  dest: {
    ext: ".html",
    path: "/extends/index",
  },
  src: {
    asset: undefined,
    ext: ".pug",
    path: "/extends",
    remote: undefined,
    slug: "extends",
  },
}
`;

snapshot[`build a site with eta 5`] = `
{
  content: '<html class="no-js" lang="en"><head><meta charset="utf-8"/><title>Markdown content</title></head><body><h1>This is a title</h1>
<p>This is a paragraph</p>
<ul>
<li>Option 1</li>
<li>Option 2</li>
</ul></body></html>',
  data: {
    content: "extends ./_includes/layout.pug
block content
  :md
    # This is a title

    This is a paragraph

    - Option 1
    - Option 2
",
    date: 1970-01-01T00:00:00.000Z,
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    slug: "filter",
    tags: [
    ],
    title: "Markdown content",
    url: "/article.html",
  },
  dest: {
    ext: ".html",
    path: "/article",
  },
  src: {
    asset: undefined,
    ext: ".pug",
    path: "/filter",
    remote: undefined,
    slug: "filter",
  },
}
`;

snapshot[`build a site with eta 6`] = `
{
  content: '<html class="no-js" lang="en"><head><meta charset="utf-8"/><title>Pug example</title></head><body><header><h1>Pug example</h1></header></body></html>',
  data: {
    content: "header
  h1 #{title}
",
    date: 1970-01-01T00:00:00.000Z,
    layout: "layout2.pug",
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    slug: "layout",
    tags: [
    ],
    title: "Pug example",
    url: "/layout/",
  },
  dest: {
    ext: ".html",
    path: "/layout/index",
  },
  src: {
    asset: undefined,
    ext: ".pug",
    path: "/layout",
    remote: undefined,
    slug: "layout",
  },
}
`;
